linkerd/app/gateway/src/gateway.rs

```diff
@@ -0,0 +1,60 @@
+use futures::{future, Future, Poll};
+use linkerd2_app_core::proxy::{http, identity};
+use linkerd2_app_core::NameAddr;
+use std::net::SocketAddr;
+
+#[derive(Clone, Debug)]
+pub(crate) enum Gateway<O> {
+    Forbidden,
+    Outbound {
+        // Source-metadata is available via request extensions set by the
+        // inbound, but we
+        source_identity: identity::Name,
+        dst_name: NameAddr,
+        dst_addr: SocketAddr,
+        outbound: O,
+    },
+}
+
+impl<O> tower::Service<http::Request<http::glue::HttpBody>> for Gateway<O>
+where
+    O: tower::Service<
+        http::Request<http::glue::HttpBody>,
+        Response = http::Response<http::boxed::Payload>,
+    >,
+    O::Error: Send + 'static,
+    O::Future: Send + 'static,
+{
+    type Response = O::Response;
+    type Error = O::Error;
+    type Future = Box<dyn Future<Item = Self::Response, Error = Self::Error> + Send + 'static>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        match self {
+            Self::Forbidden => Ok(().into()),
+            Self::Outbound { outbound, .. } => outbound.poll_ready(),
+        }
+    }
+
+    fn call(&mut self, request: http::Request<http::glue::HttpBody>) -> Self::Future {
+        match self {
+            Self::Forbidden => {
+                let rsp = http::Response::builder()
+                    .status(http::StatusCode::FORBIDDEN)
+                    .body(Default::default())
+                    .unwrap();
+                Box::new(future::ok(rsp))
+            }
+
+            Self::Outbound {
+                outbound,
+                //source_identity,
+                ..
+            } => {
+                // let headers = request.headers_mut();
+                // headers.get_all(http::header::FORWARDED)
```

Comment:
?

Reply:
this is for the followup PR... just a note to myself

Reply:
ah, gotcha, carry on then!